### ISSUE TITLE
TBB/paralle usage example

### ISSUE THREAD

**Issue by LucaCerina** [resibots/limbo#281]
Hello everyone,

I have an architecture on which i want to exploit multithreading options to do experiments with Limbo. I read that it currently does not support OpenMP, but it is unclear to me which options should be enabled to get multicore performance using TBB.

Is it sufficient the portion of code in 'limbo/test_optimizers.cpp' that enables tbb and opt_parallelrepeater available? Or do I need modifications also inside the Eval structure. A little guide on that in the documentation would be awesome.

Thank you.

**Comment by costashatz** [comment]
Hello,

Thanks for using our code!

> Is it sufficient the portion of code in 'limbo/test_optimizers.cpp' that enables tbb and opt_parallelrepeater available?

The only prerequisite is to install libtbb (libtbb-dev package in ubuntu).. Then, it depends on the usage you want to do. There are 2 ways to use tbb inside limbo:

- Use an optimizer that supports parallel evaluations; see the [optimization tutorial](http://www.resibots.eu/limbo/tutorials/opt.html)
- Use it in your own code (i.e., inside your Eval function or wherever you need it); you can use the following:

```cpp
#include <limbo/tools/parallel.hpp>

limbo::tools::par::loop(start, end, function_pointer);
```

where `function_pointer` is either a functor or a lambda. This should have the following structure:

**functor**

```cpp
struct functor {
    return_type operator()(size_t i) {}
};
```

or

**lambda**

```cpp
[&](size_t i) {}
```

Note that if you want to use Bayesian optimization with limbo, we currently only support one-by-one evaluations; i.e., we make one evaluation and then decide what to try next. Of course, your evaluation function could have parallel calls to other things inside.

I hope I was clear and helped you!

**Comment by LucaCerina** [comment]
Thank you it was really helpful. 

For example, could I use Limbo to parallelize the random sampling initial portion in which samples are independent one from another?

**Comment by jbmouret** [comment]
We did not implement it but it can be done easily by creating a new initializer. Typically:
https://github.com/resibots/limbo/blob/master/src/limbo/init/random_sampling.hpp 
-> replace the for loop by a parallel loop (```limbo::tools::par::loop(0, _dim_out, [&](size_t i) { /* content*/});```, store the results, and add them to the model (add_sample) once the loop is done.

**Comment by LucaCerina** [comment]
It worked like a charm, thank you!

**Comment by costashatz** [comment]
Great! Feel free to open another issue if you have any question or comment..
